### PR TITLE
LOOP-4017 Fix picker bouncing

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-4017

### CODE REVIEW

File: LoopKitUI/Views/ResizeablePicker.swift [hunk 1]
```diff
@@ -43,33 +43,41 @@ struct ResizeablePicker<SelectionValue>: UIViewRepresentable where SelectionValu
...
+        view.reloadAllComponents()
         if view.selectedRow(inComponent: 0) != selectedRow {
-            view.selectRow(selectedRow, inComponent: 0, animated: true)
+            view.selectRow(selectedRow, inComponent: 0, animated: false)
         }
     }
 
     class Coordinator: NSObject, UIPickerViewDataSource, UIPickerViewDelegate {
         private var picker: ResizeablePicker
+        private var data: [SelectionValue]
```
Comment (rickpasetto) [inline LoopKitUI/Views/ResizeablePicker.swift:55]: [nit] not exactly sure why this was necessary to fix the bouncing...?

File: LoopKitUI/Views/ScheduleEditor.swift [hunk 2]
```diff
@@ -147,26 +147,31 @@ struct ScheduleEditor<Value: Equatable, ValueContent: View, ValuePicker: View, A
                 .navigationBarBackButtonHidden(true)
                 .navigationBarItems(leading: backButton, trailing: trailingNavigationItems)
         case .settings:
-            pageWithCancel
+            page
```
Comment (rickpasetto) [inline LoopKitUI/Views/ScheduleEditor.swift:150]: [nit] This also...how does the code in this file avoid the bouncing?

File: LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift [hunk 3]
```diff
@@ -68,20 +68,25 @@ public struct CorrectionRangeOverridesEditor: View {
...
     }
-    
-    private var contentWithCancel: some View {
-        if value == initialValue {
-            return AnyView(content
-                .navigationBarBackButtonHidden(false)
-                .navigationBarItems(leading: EmptyView())
-            )
+
+    private var shouldAddCancelButton: Bool {
```
Comment (rickpasetto) [inline LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift:72]: [nit] ditto

**Review by rickpasetto — COMMENTED** [review]
Just some questions about how some of the code appears unrelated, other than that LGTM

**Review by rickpasetto — APPROVED** [review]

File: LoopKitUI/Views/ResizeablePicker.swift (hunk 1 above)
Comment (ps2) [reply LoopKitUI/Views/ResizeablePicker.swift:55]: The old strategy for updating the list of selectable values was changing the id, which makes swiftui think there is an entirely new view here, and everything would be reconstructed. View identity loss causes a completely new UIPickerView to be created, which results in the "bouncing" animation. This update keeps the view identity intact, but still has to update the available rows as they change. This is how the new available rows are passed from the new ResizeablePicker (struct) to the underlying UIPickerView.

File: LoopKitUI/Views/ScheduleEditor.swift (hunk 2 above)
Comment (ps2) [reply LoopKitUI/Views/ScheduleEditor.swift:150]: The views here are relying on structural identity for swiftui to know when to update vs when to recreate. In the previous code here, the pageWithCancel was returning views of different types when the cancel button was shown vs when it wasn't. The change here fixes the improper use of AnyView, and also creates a stable structural identity for page. Before, the entire page was losing its identity and being recreated.

File: LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift (hunk 3 above)
Comment (ps2) [reply LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift:72]: Hopefully answered with the above.

**Pull request merged**